src/main/java/com/pusher/client/channel/impl/PrivateEncryptedChannelImpl.java

```diff
@@ -0,0 +1,115 @@
+package com.pusher.client.channel.impl;
+
+import com.pusher.client.AuthorizationFailureException;
+import com.pusher.client.Authorizer;
+import com.pusher.client.channel.ChannelState;
+import com.pusher.client.channel.PrivateEncryptedChannel;
+import com.pusher.client.channel.PrivateEncryptedChannelEventListener;
+import com.pusher.client.channel.SubscriptionEventListener;
+import com.pusher.client.connection.impl.InternalConnection;
+import com.pusher.client.crypto.nacl.SecretBoxOpener;
+import com.pusher.client.util.Factory;
+import com.pusher.client.util.internal.Base64;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+public class PrivateEncryptedChannelImpl extends ChannelImpl implements PrivateEncryptedChannel {
+
+    private final InternalConnection connection;
+    private final Authorizer authorizer;
+    protected SecretBoxOpener secretBoxOpener;
+
+    public PrivateEncryptedChannelImpl(final InternalConnection connection,
+                                       final String channelName,
+                                       final Authorizer authorizer,
+                                       final Factory factory) {
+        super(channelName, factory);
+        this.connection = connection;
+        this.authorizer = authorizer;
+    }
+
+    @Override
+    public void bind(final String eventName, final SubscriptionEventListener listener) {
+
+        if (!(listener instanceof PrivateEncryptedChannelEventListener)) {
+            throw new IllegalArgumentException(
+                    "Only instances of PrivateEncryptedChannelEventListener can be bound " +
+                            "to a private encrypted channel");
+        }
+
+        super.bind(eventName, listener);
+    }
+
+    protected byte[] checkAuthentication() {
+
+        try {
+            final Map authResponseMap = GSON.fromJson(getAuthResponse(), Map.class);
+            final String authKey = (String) authResponseMap.get("auth");
+            final String sharedSecret = (String) authResponseMap.get("shared_secret");
+
+            if (authKey == null || sharedSecret == null) {
+                throw new AuthorizationFailureException("Didn't receive all the fields expected " +
+                        "from the Authorizer, expected an auth token and shared_secret.");
+            } else {
+                secretBoxOpener = new SecretBoxOpener(Base64.decode(sharedSecret));
+                return authKey.getBytes();
```

Comment:
let's use `auth` consistently, currently there's also `authKey` and "auth token" being used

Reply:
👍